### ISSUE TITLE
Crash because of clean_all_lines

### ISSUE THREAD

**Issue by tebtro** [4coder-editor/4coder#146]
So I have defined a command to execute `clean_all_lines` and `buffer_save`.
But then 4coder crashes somewhere in `auto_indent_buffer`.

I think the indentations array passed to set_line_indents is empty:
```
internal b32 auto_indent_buffer(Application_Links *app, Buffer_ID buffer, Range_i64 pos, Indent_Flag flags, i32 tab_width, i32 indent_width){
...
    i64 *indentations = get_indentation_array(app, scratch, buffer, line_numbers, flags, tab_width, indent_width);
    set_line_indents(app, scratch, buffer, line_numbers, indentations, flags, tab_width);
...
```

**Comment by 4coder-editor** [comment]
I have played with this a little bit and I can't reproduce the bug so far.  I did find that the implementation of `clean_all_lines` was suspiciously over-piggy and I've tightened it up, but I have no idea what as causing the issue so I'll need more info when the next build (4.1.3) comes out to see if this is still an issue and if so how to reproduce it.

**Comment by tebtro** [comment]
Thanks for looking at it.
It is still happening in 4.1.3 and I also couldn't track down what really causes the crash, but I think I found a solution that works for me.

So the problem was that when calling `clean_all_lines` before `auto_indent_buffer`, it somehow messes up the indentation and causes a crash in `auto_indent_buffer`. If you call `auto_indent_buffer` first it works fine.

I think auto_indent_buffer somehow crashes, because in clean_all_lines trailing whitespaces are also removed from empty lines. So my fix was to check if a line only contains whitespaces and don't remove them. So trailing whitespaces only get removed from lines with text.

```
        if (has_tail_whitespace) {
            String_Const_u8 line = push_buffer_range(app, scratch, buffer, Ii64(line_start, line_end));
            if (line.size > 0) {
                {
                    b32 is_all_whitespace = true;
                    for (int i = 0; i < line.size; ++i) {
                        if (!character_is_whitespace(line.str[i])) {
                            is_all_whitespace = false;
                            break;
                        }
                    }
                    if (is_all_whitespace)  continue;
                }
```


The crash happens when saving this file for example: [https://github.com/tebtro/4coder_custom/blob/master/custom/vim.cpp](https://github.com/tebtro/4coder_custom/blob/master/custom/vim.cpp)

**Comment by 4coder-editor** [comment]
This turned out to be a tricky one. By putting clean all lines before auto indent you found a non-deterministic crash bug. Here's how it worked:
1. Clean all lines happens to decide to edit a very large range of a file because it sees spots near the top and bottom it wants to change, even though they get split into lots of little edits as a batch edit, the single edit is treated as having a single large range so that there aren't 100s or 1000s of little maintenance things done for that batch edit.
2. If the file is big enough this triggers an asynchronous relex and reparse action. If the file isn't big enough to trigger this the bug doesn't happen. If you manually trigger clean all lines the async job completes in time for there to be no problem. If you're testing this and you get lucky you get a crash related to the tokens looking corrupted while auto-indenting, but more likely you get a crash related to the fact that you're looking at the old tokens and their positions are just slightly wrong, which makes it a lot harder to determine that that's the issue.

The fix was to introduce a mechanism to tell code that needs up to date tokens to check for an async task to update the tokens and wait on it if necessary. Unfortunately this case means nullifying the advantage of the async lexer, but it prevents the crash and replaces it with a slight potential wait if the lex is taking any notable amount of time.

As another level of safety against crashes I also clamped the indentation so that it can not go negative (that was one of the possible causes of a crash in this bug).

Finally just to prevent 4coder from "overworking" I stopped the clean lines call from touching blank lines, which was causing all these problems and extra computations to happen *every time* you saved instead of just when there was actually end of line whitespace all over the buffer.

Fixes in 4.1.4.
